### PR TITLE
Use Validated in GFW Pro Dashboard and Forest Change Diagnostic

### PR DESCRIPTION
**Pull request by jpolchlo**
## Pull request checklist

Please check if your PR fulfills the following requirements:
- [x] Make sure you are requesting to pull a topic/feature/bugfix branch (right side). Don't request your master!
- [x] Make sure you are making a pull request against the develop branch (left side). Also you should start your branch off our develop.
- [x] Check the commit's or even all commits' message styles matches our requested structure.
- [ ] Check your code additions will fail neither code linting checks nor unit test.

## Pull request type

 

 

Please check the type of change your PR introduces:
- [x] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe): 

## What is the current behavior?

We have encountered bad inputs in GFW Pro Dashboard (see GP-96) which lead to topology exceptions during the intersections required for feature splitting.  These are fatal exceptions.

Issue Number: N/A

## What is the new behavior?

The solution should be to use the same strategy as in #122 to capture intersection problems in the input geometries by threading the use of `Validated` from cats into the Dashboard module.

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Other information

Note that this PR was built on top of #127 to aid in debugging.

### CODE REVIEW

File: src/main/scala/org/globalforestwatch/ValidatedWorkflow.scala [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+object ValidatedWorkflow {
+  def apply[E, A: ClassTag](rdd: RDD[Validated[E, A]]): ValidatedWorkflow[E, A] = {
+    val valid = rdd.collect {
+      case Valid(a) => a
+    }
+    val invalid = rdd.collect{
+      case row: Invalid[E] => row
+    }.repartition(math.max(1, math.log10(rdd.getNumPartitions).toInt))
```
Comment (pomadchin) [inline src/main/scala/org/globalforestwatch/ValidatedWorkflow.scala:47]: 💯

File: src/main/scala/org/globalforestwatch/summarystats/ErrorSummaryRDD.scala [hunk 2]
```diff
@@ -148,19 +149,21 @@ trait ErrorSummaryRDD extends LazyLogging with java.io.Serializable {
...
+          summary match {
+            // If there was no intersection for any partial results, we consider this an invalid geometry
+            case Valid(NoIntersection) =>
+              Invalid(Location(fid, NoIntersectionError))
+            case Valid(GTSummary(result)) if result.isEmpty =>
+              Invalid(Location(fid, NoIntersectionError))
+            case Invalid(error) =>
+              Invalid(Location(fid, error))
+            case Valid(GTSummary(result)) =>
+              Valid(Location(fid, result))
```
Comment (pomadchin) [inline src/main/scala/org/globalforestwatch/summarystats/ErrorSummaryRDD.scala:165]: Could be some `.toLocation` syntax and reused if neccesary.

File: src/main/scala/org/globalforestwatch/util/PartitionVisualization.scala [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+          Iterator.single(p)
+        } else Iterator.empty
+      }
+
+    val features = partitionFeatures.collect.map { extent =>
+      Feature(extent.toPolygon(), PartitionInfo(0))
+    }
+    JsonFeatureCollection(features).asJson
+  }
+}
```
Comment (jpolchlo) [inline src/main/scala/org/globalforestwatch/util/PartitionVisualization.scala:37]: Well, this is neat.  What does the output look like?

**Review by jpolchlo — COMMENTED** [review]
I'm not allowed to approve my own PR, but thanks for carrying the ball on this, Eugene.  Some nice looking enhancements here.  I think approval on this should be pending based on how it performs in QA on some real data; I look forward to word on that.

**Pull request merged**